pkg/infrastructure/clusterapi/clusterapi.go

```diff
@@ -362,49 +362,54 @@ func (i *InfraProvider) Provision(ctx context.Context, dir string, parents asset
 }
 
 // DestroyBootstrap destroys the temporary bootstrap resources.
-func (i *InfraProvider) DestroyBootstrap(dir string) error {
+func (i *InfraProvider) DestroyBootstrap(ctx context.Context, dir string) error {
 	metadata, err := metadata.Load(dir)
 	if err != nil {
 		return err
 	}
 
-	// TODO(padillon): start system if not running
-	if sys := clusterapi.System(); sys.State() == clusterapi.SystemStateRunning {
-		machineName := capiutils.GenerateBoostrapMachineName(metadata.InfraID)
-		machineNamespace := capiutils.Namespace
-		if err := sys.Client().Delete(context.TODO(), &clusterv1.Machine{
-			ObjectMeta: metav1.ObjectMeta{
-				Name:      machineName,
-				Namespace: machineNamespace,
-			},
-		}); err != nil {
-			return fmt.Errorf("failed to delete bootstrap machine: %w", err)
+	sys := clusterapi.System()
+	if sys.State() != clusterapi.SystemStateRunning {
+		if err := sys.Run(ctx); err != nil {
+			return fmt.Errorf("failed to run capi system: %w", err)
 		}
+	}
 
-		machineDeletionTimeout := 2 * time.Minute
-		logrus.Infof("Waiting up to %v for bootstrap machine deletion %s/%s...", machineDeletionTimeout, machineNamespace, machineName)
-		machineContext, cancel := context.WithTimeout(context.TODO(), machineDeletionTimeout)
-		wait.Until(func() {
-			err := sys.Client().Get(context.TODO(), client.ObjectKey{
-				Name:      machineName,
-				Namespace: machineNamespace,
-			}, &clusterv1.Machine{})
-			if err != nil {
-				if apierrors.IsNotFound(err) {
-					logrus.Debugf("Machine deleted: %s", machineName)
-					cancel()
-				} else {
-					logrus.Debugf("Error when deleting bootstrap machine: %s", err)
-				}
-			}
-		}, 2*time.Second, machineContext.Done())
+	machineName := capiutils.GenerateBoostrapMachineName(metadata.InfraID)
+	machineNamespace := capiutils.Namespace
+	if err := sys.Client().Delete(ctx, &clusterv1.Machine{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      machineName,
+			Namespace: machineNamespace,
+		},
+	}); err != nil {
+		return fmt.Errorf("failed to delete bootstrap machine: %w", err)
+	}
 
-		err = machineContext.Err()
-		if err != nil && !errors.Is(err, context.Canceled) {
-			logrus.Infof("Timeout deleting bootstrap machine: %s", err)
+	machineDeletionTimeout := 5 * time.Minute
+	logrus.Infof("Waiting up to %v for bootstrap machine deletion %s/%s...", machineDeletionTimeout, machineNamespace, machineName)
+	machineContext, cancel := context.WithTimeout(ctx, machineDeletionTimeout)
+	wait.Until(func() {
+		err := sys.Client().Get(ctx, client.ObjectKey{
+			Name:      machineName,
+			Namespace: machineNamespace,
+		}, &clusterv1.Machine{})
+		if err != nil {
+			if apierrors.IsNotFound(err) {
+				logrus.Debugf("Machine deleted: %s", machineName)
+				cancel()
+			} else {
+				logrus.Debugf("Error when deleting bootstrap machine: %s", err)
+			}
 		}
+	}, 2*time.Second, machineContext.Done())
```

Comment:
agreed, the contexts could be cleaned up and I incorporated your changes into a fixup commit: https://github.com/openshift/installer/pull/8350/commits/7e225c354358b2894796a27a4f9dd2ee2b2a88d6

but note that a child context cannot cancel a parent.

Reply:
But you don't need to cancel the parent, you just need to stop the `wait.Until...` right? I think what you have now is worse: we don't want to cancel the context we received in the `DestroyBootstrap` hook since we could still run other things during the overall bootstrap destroy timeout. We just want to cancel the specific 2min timeout waiting for the bootstrap machine to be destroyed.

Reply:
And your point is that we don't have the `cancel()` from the parent, so we can't cancel it anyway so my initial observation "and you're cancelling the parent one" was wrong. Got it, sorry for that.